### PR TITLE
R2RDump: Add support for partially interruptible regions

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
We were basically missing this block:

https://github.com/dotnet/coreclr/blob/030a3ea9b8dbeae89c90d34441d4d9a1cf4a7de6/src/vm/gcinfodecoder.cpp#L726-L810

### CODE REVIEW

File: src/tools/r2rdump/Amd64/GcInfo.cs [hunk 1]
```diff
@@ -186,7 +187,15 @@ public GcInfo(byte[] image, int offset, Machine machine, ushort majorVersion)
...
 
-            Transitions = GetTranstions(image, ref bitOffset);
+            // Try partially interruptible first
+            if (NumSafePoints > 0)
+            {
+                LiveSlotsAtSafepoints = GetLiveSlotsAtSafepoints(image, ref bitOffset);
+            }
+            else
+            {
+                Transitions = GetTransitions(image, ref bitOffset);
```
Comment (trylek) [inline src/tools/r2rdump/Amd64/GcInfo.cs]: Thanks for fixing the typo, that kept annoying me for quite some time but I never found the energy to send out a PR just to fix a naming typo :-).

**Comment by trylek** [comment]
FYI - I have pulled your private branch and started dogfooding the GC info fix. I hit a nullref in TextDumper.cs at line 206 when dumping

<pre>
D:\git\coreclr2\bin\tests\Windows_NT.x64.Debug\JIT\Methodical\divrem\div\decimaldiv_cs_d\CPAOT-ret.out\decimaldiv_cs_d.exe
</pre>

because rtf.Method.GcInfo.Transitions was null. I have modified the line like this:

<pre>
                if (rtf.Method.GcInfo?.Transitions?.ContainsKey(codeOffset) ?? false)
</pre>

This obviously fixes the nullref but I'm not sure if it's the right fix - it seems somewhat concerning that we don't have Transitions where previously we had some.

**Comment by trylek** [comment]
With the nullref fix, I'm still seeing 32 failures when dumping some tests under

<pre>
D:\git\coreclr2\bin\tests\Windows_NT.x64.Debug\JIT\Methodical\divrem
</pre>

For instance, the test

<pre>
D:\git\coreclr2\bin\tests\Windows_NT.x64.Debug\JIT\Methodical\divrem\rem\overldrem_cs_d\Crossgen-ret.out\overldrem_cs_d.exe
</pre>

compiled with both CPAOT and Crossgen crashes with the following call stack:

<pre>
FailFast: 
   at System.Diagnostics.Debug.Assert(Boolean condition, String message, String detailMessage)
   at R2RDump.Amd64.GcSlotTable.get_NumTracked() in D:\git\coreclr\src\tools\r2rdump\Amd64\GcSlotTable.cs:line 69
   at R2RDump.Amd64.GcInfo.GetLiveSlotsAtSafepoints(Byte[] image, Int32& bitOffset) in D:\git\coreclr\src\tools\r2rdump\Amd64\GcInfo.cs:line 364
   at R2RDump.Amd64.GcInfo..ctor(Byte[] image, Int32 offset, Machine machine, UInt16 majorVersion) in D:\git\coreclr\src\tools\r2rdump\Amd64\GcInfo.cs:line 193
   at R2RDump.R2RReader.ParseRuntimeFunctions(Boolean[] isEntryPoint, Int32 runtimeFunctionOffset, Int32 runtimeFunctionSize) in D:\git\coreclr\src\tools\r2rdump\R2RReader.cs:line 558
   at R2RDump.R2RReader..ctor(DumpOptions options, String filename) in D:\git\coreclr\src\tools\r2rdump\R2RReader.cs:line 383
   at R2RDump.R2RDump.Run(String[] args) in D:\git\coreclr\src\tools\r2rdump\R2RDump.cs:line 501
   at R2RDump.R2RDump.Main(String[] args) in D:\git\coreclr\src\tools\r2rdump\R2RDump.cs:line 575
   at System.Environment.FailFast(System.String, System.Exception)
   at R2RDump.Amd64.GcSlotTable.get_NumTracked()
   at R2RDump.Amd64.GcInfo.GetLiveSlotsAtSafepoints(Byte[], Int32 ByRef)
   at R2RDump.Amd64.GcInfo..ctor(Byte[], Int32, System.Reflection.PortableExecutable.Machine, UInt16)
   at R2RDump.R2RReader.ParseRuntimeFunctions(Boolean[], Int32, Int32)
   at R2RDump.R2RReader..ctor(R2RDump.DumpOptions, System.String)
   at R2RDump.R2RDump.Run(System.String[])
   at R2RDump.R2RDump.Main(System.String[])
</pre>

**Comment by MichalStrehovsky** [comment]
> I hit a nullref in TextDumper.cs at line 206 when dumping

Yep, I had a fix locally that I didn't bother to push out since this can't be merged for another week. Pushed it out.

> This obviously fixes the nullref but I'm not sure if it's the right fix - it seems somewhat concerning that we don't have Transitions where previously we had some.

The piece of code that I'm quoting in the first post in gcinfodecoder.cpp jumps over the part that builds "Transitions", so I think this is expected.

> compiled with both CPAOT and Crossgen crashes with the following call stack

I'll take a look at that.

**Comment by MichalStrehovsky** [comment]
>> compiled with both CPAOT and Crossgen crashes with the following call stack

> I'll take a look at that.

Pushed out a fix. We were not actually decoding all the stack slots because there was an artificial limit. This codebase needs more asserts.

**Comment by trylek** [comment]
Wonderful, thank you! I can confirm that with this additional delta I no longer see any R2RDump failures in the JIT\Methodical\divrem category. If by any chance you came to the conclusion that you started liking work on R2RDump and that you don't have anything better to do (I understand it's not too likely), it would be wonderful if you could take a look why R2RDump doesn't publish coredistools.dll to the netcoreapp2.0 folder along with the executable, that's the only other super annoying problem with R2RDump I'm aware of today after your GC info fix.

**Comment by mjsabby** [comment]
I now get an IndexOutOfRangeException on one of my assemblies. I've shared it offline with both of you. Where can I find coredistools.dll?

**Comment by MichalStrehovsky** [comment]
> I now get an IndexOutOfRangeException on one of my assemblies. I've shared it offline with both of you. Where can I find coredistools.dll?

The `IndexOutOfRangeException` was the thing this pull request was fixing - I don't see it for your binary with the latest master.

**Comment by mjsabby** [comment]
Yup, it does fix it.

**Pull request merged**